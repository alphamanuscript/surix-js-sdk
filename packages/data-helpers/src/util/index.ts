export {
  deflateValue,
  makeField,
  toFieldPathArray,
  walkEntityPath,
  walkFieldPath,
  expandQuery
} from './util';