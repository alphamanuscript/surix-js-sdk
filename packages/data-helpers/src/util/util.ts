import {
  ApiEntity,
  Field,
  FieldPathArray,
  FieldType,
  FieldValue,
  ListValue,
  ObjectValue, 
  TypeValuePair} from '../types';

type WalkFieldArg = TypeValuePair | Field;
type WalkFieldReturn = TypeValuePair | Field | undefined;

export function toFieldPathArray (fieldPath: string): FieldPathArray {
  return fieldPath ? fieldPath.split('.') : [];
}

export function walkEntityPath (entity: ApiEntity, path: FieldPathArray): WalkFieldReturn {
  const [firstKey, ...pathRemaining] = path;
  if (firstKey in entity.data) {
    return walkFieldPath(entity.data[firstKey], pathRemaining);
  }
}

export function walkFieldPath (field: WalkFieldArg, path: FieldPathArray): WalkFieldReturn {
  const [firstKey, ...restPath] = path;
  if (typeof firstKey === 'undefined') {
    return field;
  }
  if (field.type === 'object' || field.type === 'list') {
    const value = field.value as ObjectValue | ListValue;
    if (firstKey in value) {
      const subField = value[firstKey];
      return walkFieldPath(subField, restPath);
    }
  }
}

export function deflateValue (value: any, type: FieldType): any {
  switch (type) {
    case 'datetime':
      return new Date(value);
    case 'number':
      return Number(value);
    case 'list':
      return deflateList(value);
    case 'object':
      return deflateObject(value);
    case 'boolean':
    case 'text':
    case 'file':
    default:
      return value;
  }
}

export function deflateText (textValue: string): string {
  return textValue;
}

export function deflateNumber (numValue: string): number {
  return Number(numValue);
}

export function deflateObject (obj: ObjectValue): object {
  return Object.keys(obj)
    .reduce((result, key) => {
      const { type, value } = obj[key];
      return {
        ...result,
        [key]: deflateValue(value, type)
      };
    }, {});
}

export function deflateList (list: ListValue): any[] {
  return list.map(({ type, value }) => deflateValue(value, type));
}

export function makeField (value: FieldValue, type: FieldType, name: string, label: string): Field {
  return {
    value,
    type,
    name,
    label
  };
}