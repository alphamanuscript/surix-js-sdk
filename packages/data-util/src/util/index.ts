export {
  dehydrateValue,
  makeField,
  toFieldPathArray,
  walkEntityPath,
  walkFieldPath
} from './util';