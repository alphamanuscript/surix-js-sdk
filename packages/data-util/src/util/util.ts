import {
  ApiEntity,
  Field,
  FieldPathArray,
  FieldType,
  FieldValue,
  ListValue,
  ObjectValue } from '../types';

export function toFieldPathArray (fieldPath: string): FieldPathArray {
  return fieldPath.split('.');
}

export function walkEntityPath (entity: ApiEntity, path: FieldPathArray): any {
  return walkObjectPath(entity.data, path);
}

export function walkObjectPath (obj: ObjectValue | ListValue, path: FieldPathArray): any {
  const [firstKey, ...restPath] = path;
  if (typeof firstKey === 'undefined') {
    return Array.isArray(obj) ? dehydrateList(obj) : dehydrateObject(obj);
  }
  if (firstKey in obj) {
    const field = obj[firstKey];
    if (field.type === 'object' || field.type === 'list') {
      return walkObjectPath(field.value, restPath);
    }
    return dehydrateValue(field.value, field.type);
  }
}

export function dehydrateValue (value: any, type: FieldType): any {
  switch (type) {
    case 'datetime':
      return new Date(value);
    case 'number':
      return Number(value);
    case 'list':
      return dehydrateList(value);
    case 'object':
      return dehydrateObject(value);
    case 'boolean':
    case 'text':
    case 'file':
    default:
      return value;
  }
}

export function dehydrateText (textValue: string): string {
  return textValue;
}

export function dehydrateNumber (numValue: string): number {
  return Number(numValue);
}

export function dehydrateObject (obj: ObjectValue): object {
  return Object.keys(obj)
    .reduce((result, key) => {
      const { type, value } = obj[key];
      return {
        ...result,
        [key]: dehydrateValue(value, type)
      };
    }, {});
}

export function dehydrateList (list: ListValue): any[] {
  return list.map(({ type, value }) => dehydrateValue(value, type));
}

export function makeField (value: FieldValue, type: FieldType, name: string, label: string): Field {
  return {
    value,
    type,
    name,
    label
  };
}