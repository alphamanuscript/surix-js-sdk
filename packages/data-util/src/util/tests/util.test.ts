import * as util from '../util';

describe('util', () => {
  describe('makeField', () => {
    it('should create a field based on the give value and metadata', () => {
      const field = util.makeField('value', 'text', 'name', 'Label');
      expect(field).toEqual({
        name: 'name',
        label: 'Label',
        type: 'text',
        value: 'value'
      });
    });
  });
  describe('dehydrateValue', () => {
    describe('when value type is text', () => {
      it('should return the string text value', () => {
        expect(util.dehydrateValue('value', 'text')).toBe('value');
      });
    });
    describe('when value type is boolean', () => {
      it('should return the boolean value', () => {
        expect(util.dehydrateValue(true, 'boolean')).toBe(true);
      });
    });
    describe('when value type is number', () => {
      it('should the value cast to Number', () => {
        expect(util.dehydrateValue('120', 'number')).toBe(120);
      });
    });
    describe('when value type is datetime', () => {
      it('should the value cast to a Date', () => {
        const value = '2017-02-22T09:19:33.885Z';
        expect(util.dehydrateValue(value, 'datetime')).toEqual(new Date(value));
      });
    });
    describe('when value type is file', () => {
      it('should return the same file value', () => {
        const value = {
          ref: 'file21',
          mimeType: 'image/png',
          downloadUrl: '/download'
        };
        expect(util.dehydrateValue(value, 'file')).toEqual(value);
      });
    });
    describe('when value type is object', () => {
      it('should convert the value to plain object values without field metadata', () => {
        const value = {
          title: util.makeField('My Title', 'text', 'title', 'Title'),
          read: util.makeField(true, 'boolean', 'read', 'Read'),
          cover: util.makeField({
            ref: 'file1',
            mimeType: 'image/jpeg',
            downloadUrl: 'download1'
          }, 'file', 'cover', 'Cover')
        };
        expect(util.dehydrateValue(value, 'object')).toEqual({
          title: 'My Title',
          read: true,
          cover: {
            ref: 'file1',
              mimeType: 'image/jpeg',
              downloadUrl: 'download1'
          }
        });
      });
    });
    describe('when value type is list', () => {
      it('should convert value to plain array of values without field metadata', () => {
        const value = [
          { type: 'text', value: 'test' },
          { type: 'number', value: '300' },
          { type: 'datetime', value: '2017-02-22T09:19:33.885Z' },
          { type: 'file', 
            value: { ref: 'file32', mimeType: 'text/plain', downloadUrl: 'download' } }
        ];
        expect(util.dehydrateValue(value, 'list')).toEqual([
          'test', 300, new Date('2017-02-22T09:19:33.885Z'),
          { ref: 'file32', mimeType: 'text/plain', downloadUrl: 'download' }
        ]);
      });
    });
    describe('when value is complex nested object', () => {
      it('should recursively convert value to plain object without field metadata', () => {
        const value = {
          foo: util.makeField({
            bar: util.makeField({
              baz: util.makeField('abc', 'text', 'baz', 'Baz'),
              baq: util.makeField('20', 'number', 'baq', 'Baq'),
              buz: util.makeField(false, 'boolean', 'buz', 'Buz'),
              bez: util.makeField('2018-02-22T09:19:33.885Z', 'datetime', 'bez', 'Bez'),
              bat: util.makeField(
                { ref: 'f2', mimeType: 'image/png', downloadUrl: 'down1' },
                'file', 'bat', 'Bat')
            }, 'object', 'bar', 'Bar'),
          },'object', 'foo', 'Foo'),
          biz: util.makeField([
            { type: 'boolean', value: true },
            { type: 'text', value: 'xyz' },
            { type: 'datetime', value: '2018-02-24T09:20:33.885Z' },
            { type: 'number', value: '342.234' },
            { type: 'file', value: {
              ref: 'f3',
              mimeType: 'image/jpeg',
              downloadUrl: 'down3'
            } },
            { type: 'object', value: {
              fizz: util.makeField('buzz', 'text', 'fizz', 'Fizz' )
            } }
          ], 'list', 'biz', 'Biz')
        };
        expect(util.dehydrateValue(value, 'object')).toMatchSnapshot();
      });
    });
  });
});