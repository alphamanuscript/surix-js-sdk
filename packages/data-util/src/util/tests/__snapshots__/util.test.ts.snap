// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util dehydrateValue when value is complex nested object should recursively convert value to plain object without field metadata 1`] = `
Object {
  "biz": Array [
    true,
    "xyz",
    2018-02-24T09:20:33.885Z,
    342.234,
    Object {
      "downloadUrl": "down3",
      "mimeType": "image/jpeg",
      "ref": "f3",
    },
    Object {
      "fizz": "buzz",
    },
  ],
  "foo": Object {
    "bar": Object {
      "baq": 20,
      "bat": Object {
        "downloadUrl": "down1",
        "mimeType": "image/png",
        "ref": "f2",
      },
      "baz": "abc",
      "bez": 2018-02-22T09:19:33.885Z,
      "buz": false,
    },
  },
}
`;
